### DISCUSSION TITLE
simulation error occur between pause physics and unpause physics triggering uav eland

### DISCUSSION THREAD

**Discussion by bconvens** [ctu-mrs/mrs_uav_system#151] [Q&A]
This issue is a follow up of #149.

This issue also related to the simulation changes you have made. A few months ago I made and tested a code that attaches a suspended payload to a single UAV or a beam payload to 2 UAVs, but now this code gives issues. In short, I have xacro files for my loads (point mass with cable or beam with 2 cables) and I rely on your link_attacher_node. In between the load xacro launch and the rosservice call /link_attacher_node/attach_typed I (un)pause the physics via rosservice call gazebo/(un)pause_physics.

I noticed now that if I have 2 UAVs in the air and I attach only a load to a single UAV, the other drone emergency lands. The error in the gazebo tab can be reproduced by just pausing and unpausing gazebo for the time required to launch and call the service (e.g., 1-2s) and then one or both drones will eland. Before updating the code to the newest version, I could pause gazebo very long without issues. However, with the new code I noticed that attaching one beam to 2 UAVs in a single command (pause, launch, service, unpause) still works, but not in separate commands as this takes longer between pause/unpause. Also, attaching in a single command the suspended load to a single UAV, will trigger eland for the other one. So the simulation does not handle the pause like before. Any idea how to solve this? Note: I also tried to without the pause and unpause, but then my load already drops a lot due to gravity directly after I launched it and before it gets attached at the correct place. Thanks for the help!
![problem_spawn_payload](https://user-images.githubusercontent.com/12151591/212917896-4d50887f-183e-4f40-baef-815a59aaacf6.png)

_Originally posted by @bconvens in https://github.com/ctu-mrs/mrs_uav_system/discussions/149#discussioncomment-4707267_

**Comment by spurnvoj** [comment]
As I mentioned here https://github.com/ctu-mrs/mrs_uav_system/discussions/149#discussioncomment-4710206, we switched to the SDF description of the drone model. My suggestion is: 
- first, try to attach an object to the single drone to see that the connection is working and nothing is missing. 
- Then go for a connection between two drones.

**Comment by bconvens** [comment]
Hi, maybe my question was unclear. I try to summarize a bit. The connection between two drones still works and also on a single drone. The problem is when I want to connect a load to uav1 while there is also a uav2 to which I want to connect nothing at all. Then uav2 goes in eland. So the issue is purely related to the pausing of gazebo. Do you get the same issue when you pause the simulation for sufficient amount of time (e.g., 1-2s)?

**Reply by spurnvoj** [reply]
We have not tried to pause the simulation. After some tries, it looks like the lockstep is broken. https://docs.px4.io/main/en/simulation/#lockstep-simulation. So a great finding. We will try to fix this ASAP.

**Comment by spurnvoj** [comment]
Hi @bconvens, the pause of the simulation should be fixed now. Check it and let me know if you still have some problems :)

**Reply by bconvens** [reply]
Hi @spurnvoj , the pause problem is indeed fixed as now this does not cause my drones to land anymore (when I don't spawn a payload).
However, when I am spawning and attaching the payload, this still causes an issue, but as I said before only for the case of 1 load attached to 1 uav where the other uav is just hovering and then elands (case 1). However, when attaching the bar type payload to 2 uavs, this issue is not present (case2) and both drones stay in the air with the beam load.

This is what I get in case 1 (failsafe mode activated, which activates elanding):
![1uav_payload_other_uav_elands](https://user-images.githubusercontent.com/12151591/215509934-9ae29f96-1b8c-46bf-aaf4-53dc0e9649d5.png)

and in case 2 (failsafe mode not activated, which prevents elanding):
![gazebo_2uav_bar](https://user-images.githubusercontent.com/12151591/215509967-276b6115-d92c-423d-89d6-c08bfd553512.png)

Any idea why that occurs in case 1 and not in case 2?
Thanks!

**Reply by spurnvoj** [reply]
I deactivated the failsafe in pixhawk. Pull the simulation repository and try to run it again.

**Reply by bconvens** [reply]
@spurnvoj, I pulled the newest version of the mrs_uav_system today and it seems not to help anything. The same problem occurs where the UAV to which no payload was attached elands.

**Reply by klaxalk** [reply]
@bconvens could you please provide the output of the control tab of the drone that land?

Moreover, is there a possibility you could share you code and environment, so we could replicate it?
